File: templates/visualbasic/app-mvvm/ViewLocator.vb
Imports System
Imports Avalonia.Controls
Imports Avalonia.Controls.Templates
Imports AvaloniaAppTemplate.ViewModels

Public Class ViewLocator
	Implements IDataTemplate

#If AvaloniaStableChosen Then
	Public Function Build(data As Object) As Control Implements IDataTemplate.Build
#Else
	Public Function Build(data As Object) As IControl Implements IDataTemplate.Build
#End If
		If data Is Nothing Then
			Return Nothing
		End If

		Dim name = data.GetType().FullName.Replace("ViewModel", "View", StringComparison.Ordinal)
		Dim type = System.Type.GetType(name)

		If type IsNot Nothing Then
			Dim control = CType(Activator.CreateInstance(type), Control)
			control.DataContext = data
			Return control
		End If

		Return New TextBlock With {.Text = "Not Found: " & name}
	End Function

	Public Function Match(data As Object) As Boolean Implements IDataTemplate.Match
		Return TypeOf data Is ViewModelBase
	End Function
End Class

File: templates/csharp/app-mvvm/ViewLocator.cs
using System;
using Avalonia.Controls;
using Avalonia.Controls.Templates;
using AvaloniaAppTemplate.ViewModels;

namespace AvaloniaAppTemplate;

public class ViewLocator : IDataTemplate
{
#if (AvaloniaStableChosen)
    public Control? Build(object? data)
#else
    public IControl Build(object data)
#endif
    {
        if (data is null)
            return null;
        
        var name = data.GetType().FullName!.Replace("ViewModel", "View", StringComparison.Ordinal);
        var type = Type.GetType(name);

        if (type != null)
        {
            var control = (Control)Activator.CreateInstance(type)!;
            control.DataContext = data;
            return control;
        }
        
        return new TextBlock { Text = "Not Found: " + name };
    }

    public bool Match(object? data)
    {
        return data is ViewModelBase;
    }
}
